File: Full Stack Develop/Programming Language/Python/demos/d01_types.py
"""01 Types & variables — Python is dynamically typed, but supports and encourages type hints.

Key points:
  1. Variables can be used without declaring a type; the annotation `x: int = 1` is for humans and type checkers (mypy/pyright).
  2. Basic types: int / float / str / bool / bytes / None.
  3. int is arbitrary precision and never overflows.
  4. Annotations are not enforced at runtime (not checked); static checkers back them up.
  5. f-strings are the preferred way to format strings.
"""

from .log import note, show, title


def run() -> None:
    title("01 Types & variables")

    note("variables are assigned directly; annotations are optional metadata (not enforced at runtime)")
    name: str = "Harry"
    age: int = 30
    price: float = 9.99
    is_vip: bool = True
    show("name / age / price / is_vip", f"{name} / {age} / {price} / {is_vip}")

    note("int is arbitrary precision and never overflows")
    big = 2**100
    show("2 ** 100", big)

    note("dynamic typing: a variable can be rebound to a value of a different type (not recommended)")
    x: object = 42
    show("x = 42, type", type(x).__name__)
    x = "now a string"
    show("x = 'str', type", type(x).__name__)

    note("None is the only empty value; test it with is, not ==")
    maybe: str | None = None
    show("maybe is None", maybe is None)
    show("maybe or 'default'", maybe or "default")

    note("type conversions are explicit function calls")
    show('int("42")', int("42"))
    show('float("3.14")', float("3.14"))
    show("str(True)", str(True))
    show('bool("")  # empty string is falsy', bool(""))

    note("isinstance / type do runtime type checks")
    show("isinstance(1, int)", isinstance(1, int))
    show("isinstance(True, int)  # bool is a subclass of int", isinstance(True, int))

    note("f-strings support expressions and format specifiers")
    ratio = 0.8137
    show("f'{ratio:.1%}'", f"{ratio:.1%}")
    show("f'{age:>5}'", f"{age:>5}")

File: Full Stack Develop/Programming Language/Rust/src/demos/d01_types.rs
use crate::log::{note, show, title};

/// Types & variables — Rust is statically and strongly typed, inferred at compile time, zero runtime overhead.
/// Key points:
///   1. let declarations are "immutable" by default; to mutate you write let mut.
///   2. Skip the annotation when the type can be inferred; scalar types i32/i64/f64/bool/char have clear boundaries.
///   3. Integer overflow: panics in debug, handle explicitly with checked_/wrapping_.
///   4. No implicit numeric conversion; always convert explicitly with as.
///   5. Shadowing: rebind the same name with let, and you can even change its type.
pub fn run() {
    title("01 Types & variables");

    note("let is immutable by default; only mut can change");
    let name = "Harry"; // &str, inferred
    let age: i32 = 30; // explicit annotation
    let price = 9.99_f64;
    let is_vip = true;
    show("name / age / price / is_vip", format!("{name} / {age} / {price} / {is_vip}"));

    note("mutability must be explicit with mut");
    let mut counter = 0;
    counter += 1;
    show("counter after += 1", counter);

    note("shadowing: rebind the same name with let, can change type");
    let spaces = "   ";
    let spaces = spaces.len(); // now a usize
    show("shadowed spaces → len", spaces);

    note("handle integer overflow explicitly: checked returns Option, wrapping wraps around");
    let max = i32::MAX;
    show("i32::MAX", max);
    show("max.checked_add(1)", max.checked_add(1)); // None
    show("max.wrapping_add(1)", max.wrapping_add(1)); // wraps around to i32::MIN

    note("no implicit conversion; convert explicitly with as");
    let i = 7_i32;
    let f = 2.0_f64;
    show("i as f64 / f", i as f64 / f);

    note("char is a 4-byte Unicode scalar, not a byte");
    let c = '语';
    show("'语' as u32", c as u32);
    show("\"Go语言\".chars().count()", "Go语言".chars().count());
    show("\"Go语言\".len() (byte count)", "Go语言".len());

    note("const is a compile-time constant and must be annotated");
    const PI: f64 = 3.14159;
    show("PI", PI);

    note("tuple: fixed length, can hold different types, access with .0 .1");
    let pair: (i32, &str) = (1, "one");
    show("pair", pair);
    show("pair.1", pair.1);
}
